### ISSUE TITLE
Implementation of C++ Context class is not copy/move-safe

### ISSUE THREAD

**Issue by bangerth** [LLNL/sundials#113]
The page at https://github.com/LLNL/sundials/blob/develop/doc/shared/sundials/SUNContext.rst documents that there is a convenience class for C++ programmers that is implemented as follows:
```
namespace sundials
{

class Context
{
public:
   Context(void* comm = NULL)
   {
      SUNContext_Create(comm, &sunctx_);
   }

   operator SUNContext() { return sunctx_; }

   ~Context()
   {
      SUNContext_Free(&sunctx_);
   }

private:
   SUNContext sunctx_;

};

}
```

This isn't right: Since the class allocates and destroys a resource, it leads to errors if you either call the copy constructor or the copy operator, along with the corresponding move constructors and operators. A correct implementation would be as follows:
```
class Context
{
public:
   Context(void* comm = NULL)
   {
      SUNContext_Create(comm, &sunctx_);
   }

  // disallow copy and move construction
  Context(const Context &) = delete;
  Context(Context &&) = delete;

  // disallow copy and move operators
  Context & operator=(const Context &) = delete;
  Context & operator=(Context &&) = delete;

   operator SUNContext() { return sunctx_; }

   ~Context()
   {
      SUNContext_Free(&sunctx_);
   }

private:
   SUNContext sunctx_;
};
```

A better choice that allows move construction and move operators would be as follows:
```
class Context
{
public:
   Context(void* comm = NULL)
   :  sunctx_ (std::make_unique<SUNContext>())
   {
      SUNContext_Create(comm, sunctx_.get());
   }

  // disallow copy, but allow move construction
  Context(const Context &) = delete;
  Context(Context &&) = default;

  // disallow copy, but allow move operators
  Context & operator=(const Context &) = delete;
  Context & operator=(Context &&) = default;

   operator SUNContext() { return *sunctx_.get(); }

   ~Context()
   {
      if (sunctx_)
        SUNContext_Free(sunctx_.get());
   }

private:
   std::unique_ptr<SUNContext> sunctx_;
};
```
If one so chose, one could remove all of these `=delete` and `=default` declarations -- the compiler will create them automatically based on the fact that the `std::unique_ptr` member has these properties itself.

An even better implementation would also mark the single constructor as `explicit` so that accidental conversion from arbitrary pointers, in particular `nullptr` are not possible.

**Comment by balos1** [comment]
Hi @bangerth,

Thank you for pointing out the incompleteness of the class. We will fix it for the next release.

**Comment by balos1** [comment]
Fixed in 6.2.0.
